src/expr/src/scalar/mod.rs

```diff
@@ -766,6 +766,103 @@ impl MirScalarExpr {
                                     e.typ(&relation_type).scalar_type,
                                 ),
                             };
+                        } else if *func == VariadicFunc::ListIndex && is_list_create_call(&exprs[0])
+                        {
+                            // We are looking for list_index(list_create, literal), and eliminate
+                            // both the list_index and the list_create. E.g.: LIST[f1,f2][2] --> f2
+                            // We handle multi-dimensional lists as well, e.g.
+                            // LIST[[[f1, f2], [f3, f4]], [[f5, f6], [f7, f8]]] [2][n][2] -->
+                            // LIST[f6, f8] [n]
+                            // We iterate over the index_exprs and remove literals, but keep
+                            // non-literals. When we encounter a non-literal, we need to dig into
+                            // the nested ListCreates: list_creates will contain all the
+                            // ListCreates of the current level. If an element of list_creates
+                            // is not actually a ListCreate, then we break out of the loop.
+                            // When we remove a literal, we need to partial-evaluate all ListCreates
+                            // that are at the current level (except those that disappeared due to
+                            // literals at earlier levels).
+                            let index_exprs = &mut exprs.split_off(1);
+                            let top_level_list_create = &mut exprs[0];
+                            let mut list_creates = vec![top_level_list_create];
```

Comment:
I wonder if you could pull apart the `MirScalarExpr::CallVariadic`s and store their contents (`func` and `exprs`) here instead. And then put them back together at the end. This would free you from having to unwrap them every time during the optimization, which could be more readable.

Reply:
I tried to change it, but there would be still two "pulling aparts", because in the else branch of the `is_literal_ok` if I have to then do the pulling apart again.

Reply:
I guess it also doesn't work because you are replacing the elements in `list_creates` by list elements, which can be arbitrary `MirScalarExpr`s.